otherlibs/systhreads/st_win32.h

```diff
@@ -193,20 +251,41 @@ Caml_inline DWORD st_mutex_lock(st_mutex m)
 
 Caml_inline DWORD st_mutex_trylock(st_mutex m)
 {
+  DWORD rc;
+
   TRACE1("st_mutex_trylock", m);
-  if (TryEnterCriticalSection(m)) {
+  EnterCriticalSection(&m->lock);
+  if (m->status == UNLOCKED) {
+    m->status = LOCKED;
+    rc = PREVIOUSLY_UNLOCKED;
     TRACE1("st_mutex_trylock (success)", m);
-    return PREVIOUSLY_UNLOCKED;
   } else {
+    rc = ALREADY_LOCKED;
     TRACE1("st_mutex_trylock (failure)", m);
-    return ALREADY_LOCKED;
   }
+  LeaveCriticalSection(&m->lock);
+  return rc;
 }
 
 Caml_inline DWORD st_mutex_unlock(st_mutex m)
 {
+  DWORD rc = 0;
+  struct st_wait_list * curr, * next;
+
   TRACE1("st_mutex_unlock", m);
-  LeaveCriticalSection(m);
+  EnterCriticalSection(&m->lock);
+  /* Wake up all waiting threads */
+  curr = m->waiters;
+  while (curr != NULL) {
+    next = curr->next;
+    TRACE1("st_mutex_unlock: waking up", curr->event);
+    if (! SetEvent(curr->event)) rc = GetLastError();
+    curr = next;
```

Comment:
I know it's what was there before, but what's wrong with:
```suggestion
    curr = curr->next;
```
and not have `next` at all?